collab2/foraging/toolkit/proximity.py

```diff
@@ -1,93 +1,209 @@
-import math
+import copy
+from typing import Callable, List, Optional, Union
 
 import numpy as np
 import pandas as pd
 
-
-def proximity_score(distance, getting_worse=1.5, optimal=4, proximity_decay=1):
-    if distance <= getting_worse:
-        return math.sin(math.pi / (2 * getting_worse) * (distance + 3 * getting_worse))
-    elif distance <= getting_worse + 1.5 * (optimal - getting_worse):
-        return math.sin(
-            math.pi / (2 * (optimal - getting_worse)) * (distance - getting_worse)
-        )
-    else:
-        return math.sin(
-            math.pi
-            / (2 * (optimal - getting_worse))
-            * (1.5 * (optimal - getting_worse))
-        ) * math.exp(1) ** (
-            -proximity_decay * (distance - optimal - 0.5 * (optimal - getting_worse))
-        )
-
-
-def foragers_to_forager_distances(foragers):
-    forager_distances = []
-
-    for b in range(len(foragers)):
-        frames = len(foragers[0])
-        forager_distances.append(np.zeros((frames, len(foragers))))
-
-        for o in range(len(foragers)):
-            for t in range(frames):
-                forager_distances[b][t, o] = math.sqrt(
-                    (foragers[b]["x"].iloc[t] - foragers[o]["x"].iloc[t]) ** 2
-                    + (foragers[b]["y"].iloc[t] - foragers[o]["y"].iloc[t]) ** 2
-                )
-
-        forager_distances[b] = pd.DataFrame(forager_distances[b])
-        forager_distances[b].columns = range(1, len(foragers) + 1)
-
-    return forager_distances
+from collab2.foraging.toolkit import filter_by_distance
+from collab2.foraging.toolkit.utils import dataObject  # noqa: F401
 
 
-def generate_proximity_score(
-    foragers,
-    visibility,
-    visibility_range,
-    getting_worse=1.5,
-    optimal=4,
-    proximity_decay=1,
-    start=0,
-    end=None,
-    time_shift=0,
+def _piecewise_proximity_function(
+    distance: Union[float, np.ndarray],
+    getting_worse: float = 1.5,
```

Comment:
can we rename `getting_worse` to `repulsion_radius` so that its meaning is more clear?

Reply:
Relatedly, maybe we can call `optimal` as `optimal_distance`

Reply:
sure